armory/data/datasets.py

```diff
@@ -20,70 +21,98 @@
 logger = logging.getLogger(__name__)
 
 
-def _in_memory_dataset_tfds(
-    dataset_name: str, dataset_dir: str, preprocessing_fn: Callable,
-) -> (np.ndarray, np.ndarray, np.ndarray, np.ndarray):
-    default_graph = tf.compat.v1.keras.backend.get_session().graph
+class ArmoryDataGenerator(DataGenerator):
+    def __init__(self, generator, size, batch_size, preprocessing_fn=None):
+        super().__init__(size, batch_size)
+        self.preprocessing_fn = preprocessing_fn
+        self.generator = generator
+        self.total_iterations = size // batch_size
 
-    train_x, train_y = tfds.load(
-        dataset_name,
-        batch_size=-1,
-        split="train",
-        as_supervised=True,
-        data_dir=dataset_dir,
-    )
+    def get_batch(self) -> (np.ndarray, np.ndarray):
+        x, y = next(self.generator)
+        if self.preprocessing_fn:
+            x = self.preprocessing_fn(x)
+
+        return x, y
+
+
+def _generator_from_tfds(
+    dataset_name: str,
+    split_type: str,
+    batch_size: int,
+    epochs: int,
+    dataset_dir: str,
+    preprocessing_fn: Callable,
+):
+    default_graph = tf.compat.v1.keras.backend.get_session().graph
 
-    test_x, test_y = tfds.load(
+    ds, ds_info = tfds.load(
         dataset_name,
-        batch_size=-1,
-        split="test",
+        batch_size=batch_size,
+        split=split_type,
         as_supervised=True,
         data_dir=dataset_dir,
+        with_info=True,
     )
 
-    train_x = tfds.as_numpy(train_x, graph=default_graph)
-    train_y = tfds.as_numpy(train_y, graph=default_graph)
-    test_x = tfds.as_numpy(test_x, graph=default_graph)
-    test_y = tfds.as_numpy(test_y, graph=default_graph)
+    ds = ds.repeat(epochs)
+    ds = ds.shuffle(batch_size * 10)
```

Comment:
Should we consider random seed setting for replicability? (Perhaps later)

Reply:
Yeah I think we need to add a test for this and we can go from there. Looking into it I'm confused as to the default TF2 behavior for shuffle:

https://github.com/tensorflow/tensorflow/issues/31706#issuecomment-522157766

Reply:
Actually we need to make sure TF1 and TF2 are using the same type of shuffle (again probably best in another issue/pr)

Reply:
Per discussion it will be hard to achieve determinism anyway due to model initialization. Likely not an issue